File: bindings/python/samples/stream_packet_callback_sample.py
import struct
import socket
import threading

import zlink
from sample_support import tcp_endpoint, wait_socket_monitor_event


def main():
# --8<-- [start:doc]
    port, endpoint = tcp_endpoint()

    with zlink.create_context() as ctx:
        with zlink.create_stream_socket(ctx) as server:
            with server.monitor_open(zlink.MonitorEventMask.ACCEPTED) as server_monitor:
                done = threading.Event()
                observed = {}

                def on_packet(routing_id, header, body):
                    observed["header"] = header.to_bytes()
                    observed["body"] = body.to_bytes()
                    observed["routing_id"] = routing_id
                    done.set()

                server.bind(endpoint)
                server.on_packet(on_packet)
                with socket.create_connection(("127.0.0.1", port), timeout=3.0) as client:
                    wait_socket_monitor_event(server_monitor, zlink.MonitorEventMask.ACCEPTED)
                    client.sendall(struct.pack("!HI", 0, len(b"hello-stream")) + b"hello-stream")
                    if not done.wait(3.0):
                        raise TimeoutError("stream packet callback did not receive a message")
                    if observed["header"] != b"":
                        raise AssertionError("unexpected stream packet header")
                    if observed["body"] != b"hello-stream":
                        raise AssertionError("unexpected stream packet body")
                    if not observed["routing_id"]:
                        raise AssertionError("stream packet callback expected a routing id")
                    server.send(observed["routing_id"]).message(b"hello-stream").submit()
                    reply = client.recv(64)
                    if reply != b"hello-stream":
                        raise AssertionError(f"unexpected stream packet reply: {reply!r}")
            print('[stream/packet-callback] send: "hello-stream" → recv: "hello-stream"')
# --8<-- [end:doc]


if __name__ == "__main__":
    main()

File: bindings/rust/samples/stream_packet_callback_sample.rs
//! STREAM callback sample – demonstrates STREAM socket with packet handler.
//! The STREAM socket binds as a server; a raw TCP client connects inward.

#[path = "sample_support.rs"]
mod sample_support;

use std::io::Write;
use std::net::TcpStream;
use std::sync::mpsc;
use std::time::Duration;

use zlink::{Context, SocketMonitor};

fn write_stream_packet(stream: &mut TcpStream, body: &[u8]) {
    let mut frame = Vec::with_capacity(6 + body.len());
    frame.extend_from_slice(&0u16.to_be_bytes());
    frame.extend_from_slice(&(body.len() as u32).to_be_bytes());
    frame.extend_from_slice(body);
    stream.write_all(&frame).expect("tcp write failed");
    stream.flush().expect("tcp flush failed");
}

fn main() {
    // --8<-- [start:doc]
    let ctx = Context::new().expect("context creation failed");

    let mut stream = ctx.stream_socket().expect("stream socket failed");
    let endpoint = sample_support::tcp_endpoint();
    stream.bind(&endpoint).expect("bind failed");
    let stream_mon = SocketMonitor::open(&stream).expect("stream monitor open failed");

    let (tx, rx) = mpsc::channel();

    stream
        .on_packet(move |routing_id, header, body| {
            assert!(!routing_id.as_bytes().is_empty());
            assert!(header.as_bytes().is_empty());
            let _ = tx.send(body.as_bytes().to_vec());
        })
        .expect("on_packet failed");

    let tcp_addr = endpoint.strip_prefix("tcp://").unwrap();
    let mut tcp_client = TcpStream::connect(tcp_addr).expect("tcp connect failed");
    tcp_client.set_nodelay(true).expect("set_nodelay failed");
    sample_support::wait_stream_connected(&stream_mon);
    drop(stream_mon);

    write_stream_packet(&mut tcp_client, b"hello-stream");
    let payload = rx
        .recv_timeout(Duration::from_secs(5))
        .expect("stream callback did not fire within 5s");
    assert_eq!(payload, b"hello-stream");
    let recv_str = std::str::from_utf8(&payload).unwrap();
    println!(
        "[stream/packet-callback] send: \"hello-stream\" → recv: \"{}\"",
        recv_str
    );
    // --8<-- [end:doc]
}
